rollup-guide/op-stack-tutorial.md

```diff
@@ -0,0 +1,39 @@
+# OP Stack Tutorial
+
+This tutorial provides practical instructions and code examples for developers intending to integrate EthStorage with an OP Stack testnet as a long-term DA solution.
+
+## L1 Contracts
+
+In the OP Stack, the `BatchInbox` is an L1 address to which the L2 batch submitter, or batcher, submits transaction batches in the form of L1 calldata or blobs. It is currently designed as a standard Externally Owned Account (EOA) address to reduce gas costs by avoiding the execution of any EVM code.
+
+To enable EthStorage to store transaction batches as blobs, a smart contract must be deployed to accept blob transactions from the batch submitter to replace the EOA. Subsequently, the smart contract invokes the `put(key, blob)` function of the EthStorage storage contract.
+
+<figure><img src="../.gitbook/assets/rollup-contract.jpg" alt=""><figcaption><p>The BatchInbox Contract serves as the Application Contract in the rollup integration scenario</p></figcaption></figure>
+
+An early version of the `BatchInbox` smart contract with EIP-4844 integration can be found [here](https://github.com/ethstorage/es-op-batchinbox/blob/main/src/BatchInbox.sol). The smart contract employs the blobhash of the blob as the key to put the blob in the EthStorage's key-value store, and covers the upfront storage payment using its ETH balance.
+
+The EthStorage storage contract address on the Sepolia testnet can be found [here](../information/#public-testnet-1-deployment).
+
+## L2 Batch Submitter
+
+Some essential changes need to be made for the OP Stack's batch submitter when the `BatchInbox` is a smart contract.
+
+Firstly, it will check if the `BatchInbox` address is a smart contract or an EOA. It is important to estimate gas usage before submitting the transaction to the smart contract, , rather than relying on intrinsic gas as in the EOA case. Furthermore, when interacting with a smart contract, it is crucial to implement the verification of transaction status and error handling to ensure data integrity.
```

Comment:
```suggestion
Firstly, it will check if the `BatchInbox` address is a smart contract or an EOA. It is important to estimate gas usage before submitting the transaction to the smart contract, rather than relying on intrinsic gas as in the EOA case. Furthermore, when interacting with a smart contract, it is crucial to implement the verification of transaction status and error handling to ensure data integrity.
```